include/glow/Base/Tensor.h

```diff
@@ -669,6 +681,95 @@ void dumpImpl(const Tensor *T, llvm::raw_ostream &os,
 void dumpImpl(const Tensor *T, unsigned maxNumElem);
 void dumpImpl(const Tensor *T);
 
+template <class ElemTy> class Handle;
+
+/// A class that provides ability to iterate over a Handle<ElemTy>. Since it's
+/// common to have both mutating and const iterators, this class has template
+/// parameter IsConst, which is true to create const_iterator and false
+/// otherwise.
+template <class ElemTy, bool IsConst>
+class HandleIterator
+    : public std::iterator<std::random_access_iterator_tag, ElemTy> {
+  using HandleTy = typename std::conditional_t<IsConst, const Handle<ElemTy> *,
+                                               Handle<ElemTy> *>;
+  using ElemTyRef =
+      typename std::conditional_t<IsConst, const ElemTy &, ElemTy &>;
+
+  /// At every given moment, the iterator maintains an index, which is used to
+  /// access the Handle. When moving the iterator forward, the index is
+  /// incremented. Only valid elements can be accessed.
+  /// 0 <= idx_ <= handle_->size()
+  HandleTy handle_;
+  llvm::ArrayRef<size_t> sizes_;
+  size_t idx_;
+  /// Holds true if the underlying tensor has non-trivial alignment (i.e. not 1)
+  bool isAligned_;
+
+  HandleIterator() = default;
+
+  HandleIterator(HandleTy handle) : handle_(handle) {
+    sizes_ = handle->dims();
+    isAligned_ = handle->size() < handle->actualSize();
+  }
+
+  static HandleIterator begin(HandleTy handle) {
+    auto res = HandleIterator(handle);
+    res.idx_ = 0;
+    return res;
+  }
+
+  static HandleIterator end(HandleTy handle) {
+    auto res = HandleIterator(handle);
+    res.idx_ = res.handle_->size();
+    return res;
+  }
+
+  friend class Handle<ElemTy>;
+
+public:
+  HandleIterator &operator++() {
+    if (*this != handle_->end()) {
+      idx_++;
+    }
+    return *this;
+  }
+  HandleIterator &operator--() {
+    if (idx_) {
+      idx_--;
+    }
+    return *this;
+  }
+  HandleIterator operator+(int n) const {
+    auto res = HandleIterator(handle_);
+    res.idx_ = std::max(static_cast<int>(idx_) + n, 0);
+    res.idx_ = std::min(res.idx_, res.handle_->size());
+    return res;
+  }
+  HandleIterator operator-(int n) const { return *this + (-n); }
+  operator int() const { return idx_; }
+
+  ElemTyRef operator*() {
+    if (!isAligned_) {
```

Comment:
super nit but this could be moved into the handle then the size of the iterator can be reduced and indexed access on strided handles doesn't need an interator necessarily. Up to you though

Reply:
I believe that optimizing the Iterator is not a priority. And from code readability point of view I'd rather not add more fields to Handle.